### PR TITLE
Improving scale transformation service

### PR DESCRIPTION
**Pull request by clinique**
to ensure that they match in the same order than the scale definition file as discussed
in PR openhab/openhab-docs#486

Signed-off-by: Gaël L'hopital <glhopital@gmail.com>

### CODE REVIEW

**Comment by clinique** [comment]
@martinvw here is the PR for the scale transformation service

**Comment by ThomDietrich** [comment]
Related to https://github.com/openhab/openhab-docs/pull/486

Thanks!

File: extensions/transform/org.eclipse.smarthome.transform.scale.test/src/test/java/org/eclipse/smarthome/transform/scale/internal/ScaleTransformServiceTest.java [hunk 1]
```diff
@@ -98,4 +98,16 @@ public void testTransformByScaleErrorInBounds() throws TransformationException {
...
+    @Test
+    public void testEvaluationOrder() throws TransformationException {
+        // Ensures that only first matching scale as presented in the file is taken in account
+        String evaluationorder = "scale/evaluationorder.scale";
+        // This value matches two lines of the scale file
+        String source = "12";
+
+        String transformedResponse = processor.transform(evaluationorder, source);
+        Assert.assertEquals("first", transformedResponse);
+
```
Comment (martinvw) [inline extensions/transform/org.eclipse.smarthome.transform.scale.test/src/test/java/org/eclipse/smarthome/transform/scale/internal/ScaleTransformServiceTest.java]: Maybe leave out this newline, note that it also is a notice in the (upcoming) release of the static code analysis.

File: extensions/transform/org.eclipse.smarthome.transform.scale/src/main/java/org/eclipse/smarthome/transform/scale/internal/ScaleTransformationService.java [hunk 2]
```diff
@@ -37,28 +41,50 @@
...
 
+    class OrderedProperties extends Properties {
+        private static final long serialVersionUID = 3860553217028220119L;
+        private final HashSet<Object> keys = new LinkedHashSet<Object>();
+
+        public OrderedProperties() {
+        }
+
+        public Iterable<Object> orderedKeys() {
+            return Collections.list(keys());
```
Comment (martinvw) [inline extensions/transform/org.eclipse.smarthome.transform.scale/src/main/java/org/eclipse/smarthome/transform/scale/internal/ScaleTransformationService.java]: I think that we can also return `keys` direct;ly I do not yet see the added value of the current solution:

```java
Set<Object> orderedKeys() {
   return keys;
}
```

File: extensions/transform/org.eclipse.smarthome.transform.scale/src/main/java/org/eclipse/smarthome/transform/scale/internal/ScaleTransformationService.java [hunk 3]
```diff
@@ -37,28 +41,50 @@
     /** RegEx to extract a scale definition */
     private static final Pattern LIMITS_PATTERN = Pattern.compile("(\\[|\\])(.*)\\.\\.(.*)(\\[|\\])");
 
+    class OrderedProperties extends Properties {
+        private static final long serialVersionUID = 3860553217028220119L;
+        private final HashSet<Object> keys = new LinkedHashSet<Object>();
```
Comment (martinvw) [inline extensions/transform/org.eclipse.smarthome.transform.scale/src/main/java/org/eclipse/smarthome/transform/scale/internal/ScaleTransformationService.java]: Please use the diamond operator, ie replace the second occurrence of the same generic by `<>`

File: extensions/transform/org.eclipse.smarthome.transform.scale/src/main/java/org/eclipse/smarthome/transform/scale/internal/ScaleTransformationService.java [hunk 4]
```diff
@@ -37,28 +41,50 @@
...
+        public OrderedProperties() {
+        }
+
+        public Iterable<Object> orderedKeys() {
+            return Collections.list(keys());
+        }
+
+        @Override
+        public Enumeration<Object> keys() {
+            return Collections.<Object> enumeration(keys);
```
Comment (martinvw) [inline extensions/transform/org.eclipse.smarthome.transform.scale/src/main/java/org/eclipse/smarthome/transform/scale/internal/ScaleTransformationService.java]: Please use the diamond operator, ie replace the second occurrence of the same generic by `<>`

File: extensions/transform/org.eclipse.smarthome.transform.scale/src/main/java/org/eclipse/smarthome/transform/scale/internal/ScaleTransformationService.java [hunk 5]
```diff
@@ -37,28 +41,50 @@
     /** RegEx to extract a scale definition */
     private static final Pattern LIMITS_PATTERN = Pattern.compile("(\\[|\\])(.*)\\.\\.(.*)(\\[|\\])");
 
+    class OrderedProperties extends Properties {
```
Comment (martinvw) [inline extensions/transform/org.eclipse.smarthome.transform.scale/src/main/java/org/eclipse/smarthome/transform/scale/internal/ScaleTransformationService.java]: I believe we should add the author and goal for this inner class as well.

Comment (martinvw) [reply extensions/transform/org.eclipse.smarthome.transform.scale/src/main/java/org/eclipse/smarthome/transform/scale/internal/ScaleTransformationService.java]: Should this inner class be static I see no reason to have a reference to the the outer class.

Comment (martinvw) [reply extensions/transform/org.eclipse.smarthome.transform.scale/src/main/java/org/eclipse/smarthome/transform/scale/internal/ScaleTransformationService.java]: Should we document that this implementation has limitations, eg iterating over the entryset will not be Ordered, removing elements will cause misery etc.

File: extensions/transform/org.eclipse.smarthome.transform.scale.test/src/test/java/org/eclipse/smarthome/transform/scale/internal/ScaleTransformServiceTest.java [hunk 6]
```diff
@@ -98,4 +98,16 @@ public void testTransformByScaleErrorInBounds() throws TransformationException {
 
     }
 
+    @Test
+    public void testEvaluationOrder() throws TransformationException {
+        // Ensures that only first matching scale as presented in the file is taken in account
+        String evaluationorder = "scale/evaluationorder.scale";
```
Comment (martinvw) [inline extensions/transform/org.eclipse.smarthome.transform.scale.test/src/test/java/org/eclipse/smarthome/transform/scale/internal/ScaleTransformServiceTest.java]: Should we name the variable `evaluationOrder` :-)

File: extensions/transform/org.eclipse.smarthome.transform.scale/src/main/java/org/eclipse/smarthome/transform/scale/internal/ScaleTransformationService.java [hunk 7]
```diff
@@ -37,28 +41,50 @@
...
         try {
             final BigDecimal value = new BigDecimal(source);
 
-            for (final Range range : data.keySet()) {
-                if (range.contains(value)) {
-                    return data.get(range);
-                }
+            Optional<Range> match = data.keySet().stream().filter(range -> range.contains(value)).findFirst();
+            if (match.isPresent()) {
+                return data.get(match.get());
```
[the remark this answers is not in the document]
Comment (clinique) [reply extensions/transform/org.eclipse.smarthome.transform.scale/src/main/java/org/eclipse/smarthome/transform/scale/internal/ScaleTransformationService.java]: Thanks for your review @martinvw . So much left to do on so small changes :) I still have to progress.

File: extensions/transform/org.eclipse.smarthome.transform.scale/src/main/java/org/eclipse/smarthome/transform/scale/internal/ScaleTransformationService.java (hunk 4 above)
Comment (clinique) [reply extensions/transform/org.eclipse.smarthome.transform.scale/src/main/java/org/eclipse/smarthome/transform/scale/internal/ScaleTransformationService.java]: Diamond operator does not work here. So left it as is.

Comment (martinvw) [reply extensions/transform/org.eclipse.smarthome.transform.scale/src/main/java/org/eclipse/smarthome/transform/scale/internal/ScaleTransformationService.java]: Hmm, sorry then Intellij had some other complaint about that line but its not that important :-)

**Comment by sjka** [comment]
Thanks!

**Pull request merged**